README
ABOUT
_____

Fragistics is a Quake3 statistics program.

It generates several webpages containing statistics about the games
played from the games.log generated by a Quake3:Arena server. 


INSTALLATION
____________

See INSTALL.


HELP
____

  fragistics --help


SIMPLE USE
__________

Login as normal user. Try

  fragistics --logfile foo.log

This will create the stats in

  ~/.fragistics/stats/foo/


BETTER USE
__________

If you have more than one file then I recommend the spool mode.
Simply put all your logfiles (e.g., foo.log, bar.log) into

  ~/.fragistics/spool/

and call

  fragistics --spool

This will create the stats

  ~/.fragistics/stats/foo/
  ~/.fragistics/stats/bar/


PROBLEMS
________

If fragistics crashes, it is highly because of artefacts in the
gameslogfile left over by quake3. It would be nice, if you could
provide the gameslogfile to me, because I'm currently writing a
artefact corrector for typical problems in server logs (e.g. server
restarts create unfinished incorrect log lines).




OLD SCHOOL
__________

The following text (until EOF) is the original readme.txt as found in
the v1.5-release.

       ========================= snip =========================


+--------+
|Q3Logger|
+--------+
BETA 3  version 1.3.0.0  by TroZ

About
--------------------------------------------------------------------------------
Q3Logger is a a Open Source program that will take the games.log file generated 
by a Quake 3:Arena server and generate several webpages containing statistics 
about the games played on that server. There will be one Overall Stats page, one 
Player Stats page, and one page of stats for each game played on the server. 

The main features are Game by Game and 'Lifetime' stats as well as Doom like 
Killer - Killee tables for all kills and also broken up by specific weapons. 

More info is available at http://members.xoom.com\troz_\q3logger\index.html


Installation
--------------------------------------------------------------------------------
Create an empty directory and unzip Q3Logger there.
You can then either copy your games.log file into that directory,
or modify the q3logger.cfg file and change the line that says 'logpath0=' 
to contain the pathto your Quake 3 games.log file.

Then, just double click on Q3logger and in a few seconds the statistics webpages 
will be created in the output directory.  This output is based on templates in
your q3logger's 'templates' directory.  For information on modifying the tempaltes
see the web site.


Q3Logger uses a config file to store several settings.  You can modify these
to change the way Q3Logger works.  Most of the settings are commented in the
config file or their use is clear.  If the config file is deleted, a new one
with default settings will be regenerated, however, all comments will be lost.


Q3Logger can show a image of each level played and the models used by the 
players. Because of the Quake 3 License, I cannot distribute these images. 
However, it isn't too hard to setup yourself. Here's how: 

Inside your Q3Logger directory's output directory, there is an 'images' directory.
Inside the 'images' there are two directories called
'levelshots' and 'models'

You can either take your own screenshots of the levels and models or use images
from inside the Quake 3 pak0.pk3 file.  The instructions for extracting the 
imgaes from the pak file follow.
 
To get the level shots, open the pak0.pk3 file in you quake3/baseq3 directory 
with winzip. Then extract the images in the levelshots directory in the zipfile
to the levelshots directory you created. 

To get the images of the models is a little more difficult. Inside the pak file
are a series of \models\players\* directorys. There is one for each player. In 
each of these directories is 3 or more 'icon_*.tga' files. You first must 
extract these files and convert them to jpg files. I used Paint Shop Pro 
because it has a batch conversion routine to convert them all at once. The 
files also must be renamed. You must replace the word 'icon' in the filename 
with the players name. For example, the icon images in the sarge directory 
should be renamed to sarge_blue.jpg , sarge_default.jpg , sarge_krusade.jpg , 
sarge_red.jpg and sarge_roderic.jpg . Place all the converted, renamed files in 
the models directory. 

For more help with the images, see the example at the webpage. 
http://members.xoom.com/TroZ_/q3logger/index.html


Config File
--------------------------------------------------------------------------------
Q3Logger uses a config file to store several settings.  You can modify these
to change the way Q3Logger works.  Most of the settings are commented in the
config file or their use is clear.  If the config file is deleted, a new one
with default settings will be regenerated, however, all comments will be lost.

 

History
--------------------------------------------------------------------------------
Version 1.3.0.0 - Feb 14 2000
	Ablilty to sort players by any displayable stat (including forumlas!)
	Bar charts on player stat pages.
	Player per minute stats
	#CREDIT# has been updated to the new homepage
	

Version 1.2.0.0 - Jan 30 2000
	Better 1v1 Support - Consecutive match on same map treated as same game.
	Per Map Statistics - Sums up stats for all games on that map.
	Ability to mkae your own stats! - You can now specify formulas on the 
		templates which will be evalulated
	Better handling of player playing time - time spent spectating no longer
		counted as playing


Version 1.1.0.0 - Jan 20 2000
	99% configureable output without modifying the code - template based
		This is a HUGE feature, allowing anyone to custimize the output to
		show just the stats they want!
	All time players killed and players who killed you in player stats
	List of games played for each player in player stats
	MORE!

Version 1.0.2.0 - Jan 8 2000
	Config file (color, hilite, limit, and path settings)
	Support for as many as 1000 log files
	Able to totally remove or only display All Kills Kill Table for Overall Stats
	Win and placed tracking
	Limited support for 1v1, TDM and CTF

Version 1.0.1.0 - Dec 27 1999
	Fixed all known crashes
	Fixed handling of kills for kill tables - you can now have two of the same 
		bot play in the same game and kill table will be correct
	Kill table top scores now hilited
	Overall Kill tables now generated with less html code
	Better handing of game and player time - Game Exit reason listed if known

Version 1.0.0.2  Dec 18 1999
	Fixed the crash in disconnect before connect bug
	Fixed some mishandling of player time in game (still some potential 
		problems though)
	Now Handling Mod_Suicide and Mod_Telefrag kill types - still missing
		type 21

Version 1.0.0.1 origional release  Dec 15 1999
	support for colored names
	overall and per game stats
	killer / killee tables
	level and model images inserted in to html


Thanks
--------------------------------------------------------------------------------
I would like to thank everyone who emailed me for their help in debugging by 
sending in logs that would break Q3Logger or for giving me idea or for just
sending an email of encouragement.  THANK YOU ALL!
	

--------------------------------------------------------------------------------
For more info see: http://members.xoom.com/TroZ_/q3logger/index.html
You can email me at: troz_@xoommail.com


ChangeLog
2010-02-24  Roland Kuebert <rkuebert@users.sourceforge.net>

	* AUTHORS: Added Roland Kuebert to authors list.

	* crosstable.cpp (intcomp): Compare ints via long conversion in order to
	not loose information.
	
	* game.cpp (GetKills): Fixed erroneous definition of default argument.
	
	* game.cpp (GetKills): Fixed erroneous definition of default argument.

	* gameslogparser.cpp (ParseLine): Adapted to glib changes.
		(findValueToKey): Likewise.
	* game.cpp (HandleEvent): Likewise.

	* database.cpp: Include stdlib.h.
	* game.cpp: Include stdlib.h.
	* gameslogparser.cpp: Include stdlib.h.
	* htmlfile.cpp: Include stdlib.h.
	* os.cpp: Include stdlib.h.
	* os.h: Include string.h.
	* player.cpp: Include stdlib.h.
	* settings.cpp: Include stdlib.h.
	* stats.cpp: Include stdlib.h.


October 5 2002 - Version 1.6.0; The "better live" edition.

	Fragistics is now a perl frontend to the original raw fragistics
	engine. The frontend tries to provide a more "usual" usage. It has
	an own set of options that control the call of the 'rawfragistcs'
	engine.

	The rawfragistics engine is kept in nearly all of its internal
	functionality. It does better co-operation with a frontend and
	some small bugs are fixed.
	
	The source code is splitted into more files to get a more Java'oid
	one-class-one-file style and more natural making in the future.

	Compilation and installation is prepared with automake/autoconf.


April 4 2002 - Version 1.5.1; The "resurrection" edition.
	
	I took-over the Fragistics program, as found on 
	  http://planetquake.com/fragistics/

	Version 1.5.1 is based on the linux version of the last public
	release	v1.5 of the original author TroZ in April 13, 2000.

	I added a makefile, a shell script to extract the levelshots and
	player images from original .pk3-files, convert the images to be
	used in the fragistics-generated html pages, and a simple perl
	script to convert logfiles with different time format into the
	format understood by fragistics. Additionally I reworked the
	templates to a different color and table design and set them as
	the standard templates. The default result pages look a bit more
	attractive now. IMHO. :-)


	
	____________________________________________________________
	
	from here the original change log until v1.3 of the original
	author TroZ follows
	____________________________________________________________

	
	
Feb 14 2000 - Version 1.3.0.0
	Ablilty to sort players by any displayable stat (including forumlas!)
	Bar charts on player stat pages.
	Player per minute stats
	#CREDIT# has been updated to the new homepage
	

Jan 30 2000 - Version 1.2.0.0
	Better 1v1 Support - Consecutive match on same map treated as same game.
	Per Map Statistics - Sums up stats for all games on that map.
	Ability to mkae your own stats! - You can now specify formulas on the 
		templates which will be evalulated
	Better handling of player playing time - time spent spectating no longer
		counted as playing


Jan 20 2000 - Version 1.1.0.0
	99% configureable output without modifying the code - template based
		This is a HUGE feature, allowing anyone to custimize the output to
		show just the stats they want!
	All time players killed and players who killed you in player stats
	List of games played for each player in player stats
	MORE!

Jan 8 2000 - Version 1.0.2.0
	Config file (color, hilite, limit, and path settings)
	Support for as many as 1000 log files
	Able to totally remove or only display All Kills Kill Table for Overall Stats
	Win and placed tracking
	Limited support for 1v1, TDM and CTF

Dec 27 1999 - Version 1.0.1.0
	Fixed all known crashes
	Fixed handling of kills for kill tables - you can now have two of the same 
		bot play in the same game and kill table will be correct
	Kill table top scores now hilited
	Overall Kill tables now generated with less html code
	Better handing of game and player time - Game Exit reason listed if known

Dec 18 1999 - Version 1.0.0.2
	Fixed the crash in disconnect before connect bug
	Fixed some mishandling of player time in game (still some potential 
		problems though)
	Now Handling Mod_Suicide and Mod_Telefrag kill types - still missing
		type 21

Dec 15 1999 - Version 1.0.0.1 origional release  
	support for colored names
	overall and per game stats
	killer / killee tables
	level and model images inserted in to html


src/defines.h
/* defines.h
 Fragistics  -  Game Statistics program for Quake 3 Arena
 Copyright (C) 2000   Brian Risinger

see main.cpp for license info
*/

#ifndef DEFINES_H_INCLUDED
#define DEFINES_H_INCLUDED

#include <string>

using namespace std;


//kill types - id defined - cant be changed
#define TOTAL_KILLS			0

#define SHOTGUN_KILL		1
#define GAUNTLET_KILL		2
#define MACHINEGUN_KILL		3
#define GRENADE_KILL_DIRECT	4
#define GRENADE_KILL_SPLASH	5
#define ROCKET_KILL_DIRECT	6
#define ROCKET_KILL_SPLASH	7
#define PLASMA_KILL_DIRECT	8
#define PLASMA_KILL_SPLASH	9
#define RAILGUN_KILL		10
#define LIGHTNING_KILL		11
#define BFG_KILL_DIRECT		12
#define BFG_KILL_SPLASH		13

#define SELF_WATER			14
#define SELF_SLIME			15
#define SELF_LAVA			16
#define SELF_CRUSH			17
#define TELEFRAG_KILL		18
#define SELF_FALLING		19
#define SELF_SUICIDE		20
	#define UNKNOWN_3		21
#define SELF_TRIGGER_HURT	22


#define UNKNOWN_KILL		23
#define MAX_KILLTYPE		24
//max_killtype must be last of normal kill types

//aggregate kill types - can be changed
#define AGGREGATE_KILLS_MIN	25

#define A_ROCKET_KILL		25		//6+7
#define A_GRENADE_KILL		26		//4+5
#define A_PLASMA_KILL		27		//8+9
#define A_BFG_KILL			28		//12+13

#define A_EXPLOSIVE_KILL	29		//4+5+6+7
#define A_BULLET_KILL		30		//1+3
#define A_ENERGY_KILL		31		//8+9+12+13
#define A_OTHER_KILL		32		//2+10+11+18

#define A_HITSCAN_KILL		33		//1+2+3+10   ?+11?
#define A_NONHITSCAN_KILL	34		//4+5+6+7+8+9+12+13

#define A_SUICIDE_KILL		35		//14+15+16+17+19+20+22
#define A_NONWEP_KILL		36		//14+15+16+17+18+19+20+22

#define AGGREGATE_KILLS_MAX	37
#define CROSSTABLE_MAX_KILL	37//was 14 origionally



//item types - can be changed
#define ITEM_NONE			-1
#define ITEM_UNKNOWN		0
#define ITEM_GAUNTLET		1
#define ITEM_MACHINEGUN		2
#define ITEM_SHOTGUN		3
#define ITEM_GRENADEL		4
#define ITEM_ROCKETL		5
#define ITEM_LIGHTNINGGUN	6
#define ITEM_RAILGUN		7
#define ITEM_PLASMAGUN		8
#define ITEM_BFG			9
#define ITEM_ARMORSHARD		10
#define ITEM_YELLOWARMOR	11
#define ITEM_REDARMOR		12
#define ITEM_5HEALTH		13
#define ITEM_25HEALTH		14
#define ITEM_50HEALTH		15
#define ITEM_BATTLESUIT		16
#define ITEM_FLIGHT			17
#define ITEM_HASTE			18
#define ITEM_INVISIBILITY	19
#define ITEM_MEGAHEALTH		20
#define ITEM_QUADDAMAGE		21
#define ITEM_REGENERATION	22
#define ITEM_MEDKIT			23
#define ITEM_TELEPORTER		24
#define ITEM_REDFLAG		25
#define ITEM_BLUEFLAG		26
#define ITEM_BULLETS		27
#define ITEM_SHELLS			28
#define ITEM_GRENADES		29
#define ITEM_ROCKETS		30
#define ITEM_LIGHTNING		31
#define ITEM_SLUGS			32
#define ITEM_PLASMAAMMO		33
#define ITEM_BFGAMMO		34
#define ITEM_LAST			35
//ITEM_LAST must be last


							

//events
#define EVENT_NONE			0
#define EVENT_GAMESTART		1
#define EVENT_PLAYERJOIN	2
#define EVENT_PLAYERLEAVE	3
#define EVENT_ITEMPICKUP	4
#define EVENT_KILL			5
#define EVENT_GAMEEND		6
#define EVENT_SAY			7
#define EVENT_SAY_TEAM		8
#define EVENT_PLAYERINFO	9
#define EVENT_LIMITHIT		10
#define EVENT_PLAYER_SCORE	11
#define EVENT_TEAM_SCORE	12


//Games
#define GAMETYPE_FFA		0
#define GAMETYPE_1V1		1
#define GAMETYPE_TDM		3
#define GAMETYPE_CTF		4
#define GAMETYPE_ALL		-1
#define GAMETYPE_MAX		5


#define GAME_MAX_SIMU_CLIENTS		65


#define TEAM_FFA			0
#define TEAM_RED			1
#define TEAM_BLUE			2
#define TEAM_SPECTATOR		-1
#define TEAM_MAX			3


//kill tables
#define KILLTABLE_MAXHILITE 21


#define LOOP_NONE				0
#define LOOP_ALLPLAYER_TRIM		1
#define LOOP_ALLPLAYER_FULL		2
#define LOOP_GAMEPLAYER_TRIM	3
#define LOOP_GAMEPLAYER_FULL	4
#define LOOP_MAPPLAYER_TRIM		5
#define LOOP_MAPPLAYER_FULL		6
#define LOOP_MAPS				7
#define LOOP_PLAYERGAMES		8
#define LOOP_PLAYERKILLS		9
#define LOOP_PLAYERDEATHS		10
#define LOOP_MAPGAMES			11


#define LIMIT_UNKNOWN		0
#define LIMIT_TIME			1
#define LIMIT_FRAG			2
#define LIMIT_CAPTURE		3
#define LIMIT_MAX			4


#define PLAYERLIST_SORTTYPE_NORMAL	0
#define PLAYERLIST_SORTTYPE_PERGAME	1000
#define PLAYERLIST_SORTTYPE_PERHOUR	2000
#define PLAYERLIST_SORT_GAMES		0
#define PLAYERLIST_SORT_TIME		1
#define PLAYERLIST_SORT_KILLS		2
#define PLAYERLIST_SORT_DEATHS		3
#define PLAYERLIST_SORT_SUICIDES	4
#define PLAYERLIST_SORT_FRAGS		5
#define PLAYERLIST_SORT_EFF			6
#define PLAYERLIST_SORT_ITEMS		7
#define PLAYERLIST_SORT_SCORE		8
#define PLAYERLIST_SORT_FFAWINS		9
#define PLAYERLIST_SORT_1V1WINS		10
#define PLAYERLIST_SORT_TDMWINS		11
#define PLAYERLIST_SORT_CTFWINS		12
#define PLAYERLIST_SORT_FFAPLACE	13
#define PLAYERLIST_SORT_FFAWINPLACE	14
#define PLAYERLIST_SORT_SKILL		15
#define PLAYERLIST_SORT_AVEHC		16



#define LOGTYPE_GAMESLOG			0

//database settings
//0 dont store in database
//1 this is the active file - store in database - resume parsing this file from where parsing left off last time
//2 this file is stored in the database and not parsed
//3 this file is added to the database (with all 2 files which get reparsed) then the program exits - file is marked with 2
#define DB_DONTADD			0
#define DB_ACTIVEFILE		1
#define DB_FILESTORED		2
#define DB_STOREFILE		3
#define DB_MAX				4




class Games;
class PList;
class Stats;



#endif //DEFINES_H_INCLUDED
